include/deal.II/base/task_result.h

```diff
@@ -0,0 +1,501 @@
+     * object, and then release all information still associated with the
+     * task that originally computed the result.
+     */
+    void
+    wait_and_move_result() const;
+  };
+
+
+  // ------------------------------- inline functions --------------------------
+
+
+  template <typename T>
+  inline TaskResult<T>::TaskResult(TaskResult<T> &&other) noexcept
+  {
+    // First lock the other object, then move the members of the other
+    // object and reset it. Note that we do not have to wait for
+    // the other object's task to finish (nor should we).
+    std::lock_guard<std::mutex> lock(other.mutex);
+
+    result_is_available       = other.result_is_available.load();
+    other.result_is_available = false;
+
+    task = std::move(other.task);
+    other.task.reset();
+
+    task_result = std::move(other.task_result);
+    other.task_result.reset();
+  }
+
+
+
+  template <typename T>
+  inline TaskResult<T>::~TaskResult()
+  {
+    // First ensure that there is no currently running task. As
+    // documented, we consider this an error.
+    if (result_is_available == false)
+      {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        // Check again, now under the mutex. It is possible that the status
+        // has changed between the initial 'if' and the time we obtained the
+        // lock.
+        if (result_is_available == false)
+          Assert(task.has_value() == false,
+                 ExcMessage("You cannot destroy a TaskResult object "
+                            "while it is still waiting for its associated task "
+                            "to finish. See the documentation of this class' "
+                            "destructor for more information."));
+      }
+    // Having established that there is no task still running,
+    // let the compiler clear all member variables at the end of
+    // this destructor.
+  }
+
+
+
+  template <typename T>
+  inline void
+  TaskResult<T>::operator=(const Task<T> &t)
+  {
+    // First ensure that there is no currently running task. As
+    // documented, we consider this an error.
+    if (result_is_available == false)
+      {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        // Check again, now under the mutex. It is possible that the status
+        // has changed between the initial 'if' and the time we obtained the
+        // lock.
+        if (result_is_available == false)
+          Assert(
+            task.has_value() == false,
+            ExcMessage(
+              "You cannot assign a new task to a TaskResult object "
+              "while it is still waiting for its previous task "
+              "to finish. See the documentation of this class' operator=() "
+              "for more information."));
+      }
+    // Having established that there is no previous task still running,
+    // clear previous content of this class and set the current task
+    // as the one we're waiting for:
+    clear();
```

Comment:
The code in this function above this call is duplicated in clear() except for the error message, right? Maybe that would be cleaner to call clear() here.